crates/cubecl-wgpu/src/compute/controller.rs

```diff
@@ -29,21 +28,25 @@ impl WgpuAllocController {
     ///
     /// The controller and the corresponding `Allocation`.
     pub fn init(binding: SliceBinding, buffer: wgpu::Buffer, size: usize) -> (Self, Allocation) {
-        fn buffer_ptr(buffer: &wgpu::Buffer) -> NonNull<u8> {
+        let data = {
             let data = buffer.slice(..).get_mapped_range();
-            unsafe { NonNull::new_unchecked(data.as_ptr().cast_mut()) }
-        }
+            data[0..size].to_vec().into_boxed_slice()
+        };
+        let ptr = unsafe { NonNull::new_unchecked(data.as_ptr().cast_mut()) };
 
         let allocation = Allocation {
-            ptr: buffer_ptr(&buffer),
+            ptr,
             size,
             align: wgpu::COPY_BUFFER_ALIGNMENT as usize,
         };
 
+        // We unmap the buffer and release the binding so that the same buffer can be used again.
+        buffer.unmap();
```

Comment:
The goal of using pinned memory is to not call `to_vec`, that's why we kept the binding alive, making sure that memory can't be reused.